### PR TITLE
Underilla 3.0.0: reusable core, composable generation, and streamlined releases

### PR DESCRIPTION
**Pull request by kntrel**
# Underilla 3.0.0: reusable core, composable generation, and streamlined releases

Rebuild Underilla’s generation engine around a reusable core and a dedicated Paper adapter. This separates terrain composition from server integration, makes generation behavior independently testable, and adds reference-entity copying, flexible source-world paths, and generation profiling.

## Overview

Underilla continues to combine custom-world surfaces with generated underground terrain, while replacing the tightly coupled generation implementation with explicit, composable phases.

This is a major-version change: the Java API moves to `com.kntrel.mc.underilla`, existing integrations require migration, and several configuration keys change. The build now targets Paper **26.2** and requires **Java 25**.

Publishing is simplified to GitHub Releases: pushing a tag matching `v<project.version>` publishes the plugin only when the tagged commit belongs to `main`’s history and the build and automated tests succeed.

## Internal changes

- Split the single Gradle project into `underilla-core`, a reusable Java library, and `underilla-paper`, the Paper plugin.
- Remove Paper, Bukkit, and plugin-singleton dependencies from core generation and world-reading code.
- Move Java packages and the Gradle group from `fr.formiko.mc.underilla` to `com.kntrel.mc.underilla`.
- Replace the monolithic generator/merger implementation with `WorldGenerationPlan`, which defines work after noise, surface, carvers, features, and chunk loading. Biome selection and height queries are separate operations.
- Introduce `UnderillaFactory` to assemble `SURFACE`, `ABSOLUTE`, and `NONE` strategies from supplied readers, block factories, predicates, transformations, and generation settings.
- Add `PaperGenerationPlanFactory` to translate plugin configuration into core generation plans.
- Scope generation plans, caches, biome providers, and generator registration to individual worlds.
- Reduce `UnderillaChunkGenerator` to mapping Paper callbacks onto plan phases and delegating uncovered chunks to the configured fallback generator.
- Replace `CustomBiomeSource` with `UnderillaBiomeProvider`, using Paper’s biome-provider API and core biome patching rules.
- Move terrain boundaries out of world readers into composable world masks, including fixed-height, reference-surface, inverted, and combined masks.
- Add target-surface snapshots so reference terrain can fill gaps above a lower generated surface.
- Express surface cave protection through `NoodleCavesPolicy`. Reference writes can run before carvers or be deferred until afterward, depending on biome and liquid-preservation rules.
- Recompute deferred writes when their cached batch has been evicted.
- Introduce reusable patchers for ordered operations, conditions, speculative candidates, block transformations, deferred writes, and vertical bounds.
- Add invocation-scoped values so patchers can share calculated inputs safely across nested and concurrent pipeline executions.
- Separate world access from disk storage: `WorldReader` defines the read contract, while `DiskWorldReader` handles terrain and optional entity-region files.
- Add a shared, bounded chunk cache for reader wrappers, surface boundaries, and deferred writes, alongside the region-file and biome caches.
- Decode block-state palette entries once per chunk reader and clone cached blocks before applying mutable changes.
- Move block support/replacement cleanup into the after-features phase and entity cleanup into the after-load phase for newly generated chunks.
- Read reference entities from entity-region files and insert them through Paper’s limited-region adapter after features are generated.
- Replace the global timing map with structured profiling. Correlate callbacks by world and chunk, aggregate measurements, and periodically write `metrics.json`.
- Replace custom static logging helpers with SLF4J.
- Upgrade the Java toolchain from 21 to 25, target Paper 26.2, update paperweight to `2.0.0-beta.22`, and switch KntNBT to `com.github.kntrel:KntNBT:v2.3.0`.
- Add JUnit 6 tests across both modules, including real MCA fixtures, legacy-behavior comparisons, strategy boundaries, biome composition, cleanup, entity reading, deferred writes, profiling, source-path resolution, and Paper callback delegation.
- Wire root build/check tasks to both modules. Produce the release plugin at `underilla-paper/build/libs/Underilla-<version>.jar`.
- Restrict release publishing to matching `v` tags whose commits are contained in `main`; run `./gradlew build` before creating the release.

## API changes

- Replace string-based block and biome identity with the validated `ID(namespace, value)` record. `ID.of(...)` accepts qualified identifiers and defaults unqualified identifiers to `minecraft`.
- Replace `Biome.getName()` with `Biome.id()`.
- Replace `Block.getName()` and `Block.getNameSpace()` with `Block.id()`.
- Require `Block.clone()` and remove `Block.isSolidAndSurfaceBlock()`. Surface classification now uses supplied rules instead of configuration embedded in block implementations.
- Add `BlockFactory.air()` and `BlockFactory.create(ID)` for platform-specific block construction.
- Add `BiomeData`, a mutable biome value with global position, chunk coordinates, and biome-cell coordinates.
- Extend `ChunkData` with `addEntity(EntityView)` and `entities()`. The latter defaults to an empty collection for adapters without live entities.
- Add `Entity`, exposing `id()` and `remove()` for platform-neutral entity cleanup.
- Add `EntityView`, carrying entity NBT and its data version, with defensive copies of the NBT payload.
- Add `GenerationConstants` for chunk and biome-cell dimensions.
- Replace the old `Reader` interface and abstract `WorldReader` hierarchy with a read-only `WorldReader` interface and a separate abstract `DiskWorldReader`.
- Add `WorldReader.getBiomeID(...)`; remove the previous `getBiomeName(...)` overloads, `getWorldName()`, and `getLowerBlockOfSurfaceWorldYLevel(...)` from the reader API.
- Make disk readers accept explicit region directories and optional entity-region directories. Add `withChunkCache(...)` for sharing a generation plan’s chunk cache.
- Extend `ChunkReader` with entity snapshots and `getEntities()`. Make single-argument `blockFromTag(...)` a final, cached operation backed by the new protected `decodeBlockFromTag(...)` hook.
- Remove `ChunkReader`’s implementation of the former `Reader` interface, including the convenience methods previously inherited from it.
- Replace the old `Generator` and `MergeStrategy` entry points with `UnderillaFactory.surface(...)`, `.absolute(...)`, and `.none(...)`.
- Add a factory builder for reference/underground readers, vertical and horizontal bounds, merge depth, cache capacity, biome rules, block transformations, cleanup, generation flags, profiling, surface filling, and noodle-cave policy.
- Add `WorldGenerationPlan` and `WorldGenerationPlanBuilder` for assembling lifecycle patchers, biome patching, coverage, generation flags, and height resolution. Coverage-aware `tryAfter...` methods report whether a chunk phase ran.
- Add `GenerationArea`, `ChunkCoverage`, and `GenerationFlags` to describe bounds and participation in noise, surface, carver, feature, mob, and structure generation.
- Add `Altimeter` and `SurfaceAltimeter` for reference-world height queries, plus `ProfiledAltimeter` for instrumentation.
- Add `NoodleCavesPolicy.underground()` and `.surface(predicate, restoreLiquids)` to describe whether caves may remain visible through reference terrain.
- Add the generic `Patcher<T>` contract and composition helpers: `sequence`, `iff`, `take`, `value`, and `with`.
- Add `PatcherPipeline`, `ConditionalPatcher`, and `CandidatePatcher` for ordered, conditional, and speculative patching.
- Add `PatchTimeValue` for values calculated and scoped to a pipeline invocation.
- Add `ChunkBlock`, `PerBlockChunkPatcher`, `RoutingBlockPatcher`, `BlockTransformerPatcher`, and `TransformationBlockPatcher` for inspecting, transforming, and accepting block candidates.
- Add `DeferredPatcher` and its later-phase applier, plus `VerticalBoundChunkPatcher` to constrain chunk operations.
- Add `WorldMask`, `AbsoluteWorldMask`, `ReferenceHeightWorldMask`, and `UnionWorldMask`, including mask inversion.
- Add `ReferenceWorldPatchers`, `ReferenceWorldBlockPatcher`, `ReferenceWorldTest`, `ReferenceWorldTests`, and `WorldHeightMaskPatcher` for composing reference-terrain selection and placement.
- Add `EntityCleanupPatcher` for removing selected entities and transforming survivors.
- Add `ChunkCache` and typed `TopicChunkCache` views, replacing the old publicly exposed reader-cache implementations.
- Add the profiling API: `Instrumenter`, `Recorder`, `Tracker`, `Operation`, `Stopwatch`, `Measurement`, and `ChunkProfiledPatcher`.
- Retain the existing `WorldInfo`, `HeightMapType`, vector utilities, and tag interpreter under the new package namespace.

## Config changes

- Replace `surfaceWorld.name` with `surfaceWorld.worldPath` and `surfaceWorld.dimension`.
- Replace `cavesWorld.name` with `cavesWorld.worldPath` and `cavesWorld.dimension`.
- Default both source dimensions to `minecraft:overworld`. World/dimension resolution uses `<worldPath>/dimensions/<namespace>/<dimension>/region`.
- Add optional `surfaceWorld.regionPath` and `cavesWorld.regionPath`. A non-empty direct region path overrides the corresponding world/dimension terrain path.
- Add optional `surfaceWorld.entitiesPath`. Otherwise, entity data is resolved from the configured world and dimension.
- Resolve terrain and entity paths independently. Older world layouts can use explicit `regionPath` and `entitiesPath` values.
- Update automatic generation-area detection to inspect the resolved surface-region directory.
- Replace `cache.size` with `cache.regionFiles`, defaulting to `16`, and `cache.chunks`, defaulting to `64`.
- Warn when the obsolete `cache.size` key is present and when `cache.chunks` is below the recommended minimum of `32`. Both new cache settings have a minimum value of `1`.
- Standardize block, biome, and entity values around `namespace:value` identifiers. Unqualified values default to `minecraft`; configuration values are normalized to lowercase.
- Match block and entity regular expressions against complete namespaced identifiers. Update the bundled patterns and replacement maps accordingly.
- Keep biome-tag support through `#namespace:value`. Evaluate `onlyOn`/`exceptOn` selections without expanding unrestricted selections into the entire biome registry; a populated `onlyOn` takes precedence.
- Add `minecraft:sulfur_caves` to the bundled `biomesMerging.fromCavesGeneration.onlyOn` list.
- Remove `steps.cleaningBlocks` and `steps.cleaingEntities`. Cleanup now follows generation lifecycle phases rather than separate resumable tasks.
- Remove `clean.blocks.removeUnstableBlocks` and its behavior.
- Retain `clean.blocks.enabled`, `toSupport`, and `toReplace`, applying them after feature generation. Retain entity-removal settings, applying them when new chunks load.
- Make `vanillaPopulation.enabled` control both feature and mob generation; mob generation was previously always enabled.
- Accept strategy names without case sensitivity and trim surrounding whitespace.

Existing configurations should be reviewed against the new defaults, especially source-world paths, cache settings, and removed cleanup options.

## Removed components

- The original single-module source layout and `fr.formiko.mc.underilla` package namespace.
- The production `Generator`, `Merger`, `AbsoluteMerger`, and `MergeStrategy` implementations.
- The former `Reader` abstraction and world-reader responsibilities for terrain-boundary calculation and configuration lookup.
- Core dependencies on Bukkit/Paper types, plugin configuration, and static plugin access.
- The old `CustomBiomeSource` implementation.
- The Paper cleanup classes `CleanBlocks`, `CleanBlocksTask`, `CleanEntities`, `CleanEntitiesTask`, and `FollowableProgressTask`.
- `ChunkGeneratedListener`, replaced by generation-plan handling within the generator.
- Separate cleanup-task persistence, restart handling, and completion steps.
- Unstable-block survival checks, including their direct Minecraft/CraftBukkit implementation.
- The Paper plugin’s `getEndBlockTransformer`, `setEndBlockTransformer`, and `hasEndBlockTransformer` hooks.
- The global `Generator.times` map and custom static logging helpers.
- The old HydrolienF KntNBT dependency coordinate.
- The Sonar workflow, Gradle plugin, and analysis configuration.
- Gradle-managed Paper test-server support and the `runServer` task.
- Hangar publishing configuration and its GitHub Actions publishing step.
- The Modrinth publishing plugin.
- Maven publication metadata, staging configuration, signing, JReleaser deployment, and source/Javadoc JAR publication setup.
- The Gradle `extractChangelog` and `echoLatestVersionChangelog` helper tasks.

## What's new

- **Reusable generation engine:** build and test generation plans through the core library, independently of Paper.
- **Reference-entity copying:** transfer entities from the surface world’s entity-region files into newly generated chunks.
- **Flexible source selection:** choose source worlds by path and dimension, or point directly to terrain and entity-region directories.
- **Surface gap filling:** fill the space between a higher reference surface and lower generated terrain. Enabled by default when no separate underground source is supplied, and configurable through the core API.
- **Composable terrain rules:** extend generation through custom patchers, masks, transformations, and lifecycle phases.
- **Explicit cave-protection policies:** control when reference terrain is placed around carver execution, including selective surface exposure and liquid preservation.
- **Generation diagnostics:** collect correlated chunk timings and per-operation statistics in `plugins/Underilla/metrics.json`, with periodic snapshots and a final flush on shutdown.
- **Configurable chunk caching:** tune chunk retention separately from region-file caching, reducing repeated reading and boundary calculations.
- **Release publishing gated by tests:** publish the built plugin to GitHub from a matching `v` tag only after branch-membership validation and a successful build.

### CODE REVIEW

File: underilla-paper/src/main/resources/config.yml
Comment (HydrolienF) [inline underilla-paper/src/main/resources/config.yml:1]: The chunky version configured is not available in Chunky hangar & it's not reading the right config setting from config.yml.
It look like I did the error before.
The following config works.
```yml
chunkyVersion: 1.5.3
voidWorldGeneratorVersion: 1.3.12
```

Comment (HydrolienF) [reply underilla-paper/src/main/resources/config.yml:1]: I send the fix for this file & chunkyVersion in the main build.gradle.kts.

File: .github/workflows/release.yml
Comment (HydrolienF) [inline .github/workflows/release.yml:1]: I suggest adding Hangar publish back now that you have all perms on the Hangar project I did.

File: underilla-paper/build.gradle.kts
Comment (HydrolienF) [inline underilla-paper/build.gradle.kts:1]: apiVersion should probably be with the other constant.

File: build.gradle.kts
Comment (HydrolienF) [inline build.gradle.kts:1]: I suggest to add Hangar publish back. It's easy to do with the gradle plugin `io.papermc.hangar-publish-plugin`

I configured a publish to Maven central (jreleaser, signing & publishing) so that it can be used as a dependency by other plugins. cf https://central.sonatype.com/artifact/fr.formiko.mc.underilla/underilla
I think it might be usefull to reconfigure it with the kntrel naming in path.

sonarqube is a code quality checker. Completly optional, but if you which to have it back, it can be configured as a public project for free in the main sonar web site I think.

File: README.md [hunk 1]
```diff
@@ -7,14 +7,14 @@
 [ ![download][] ][downloadLink]
 [ ![discord-shield][] ][discord-invite]
 
-[**Discord**](https://discord.gg/RPNbtRSFqG) | [**Hangar**](https://hangar.papermc.io/Hydrolien/Underilla) | [**GitHub**](https://github.com/HydrolienF/Underilla) | [**Original author Github** (Outdated)](https://github.com/Jeshuakrc/Underilla)
+[**Discord**](https://discord.gg/RPNbtRSFqG) | [**Hangar**](https://hangar.papermc.io/Hydrolien/Underilla) | [**GitHub**](https://github.com/kntrel/underilla)
```
Comment (HydrolienF) [inline README.md:10]: 👍
The Discord link is a small server where I have all the plugin I'm maintaining. If you which to join I will give you dev role there. We could also move Underilla question/help channel to it's own Discord.

File: src/main/java/fr/formiko/mc/underilla/paper/cleaning/CleanBlocks.java
Comment (HydrolienF) [inline src/main/java/fr/formiko/mc/underilla/paper/cleaning/CleanBlocks.java:1]: removeUnstableBlock should be kept. It make sure there is no uncorrect block in the final world.
For example half door, flowers over an air block, grass block covered by an other blocks etc

File: build.gradle.kts
Comment (kntrel) [reply build.gradle.kts:1]: Added Hangar publishing back in [bd7975e](https://github.com/kntrel/Underilla/pull/18/commits/bd7975ef662d050090a855065ea71dbe3c028829).

Code quality control is welcomed. I removed it because it pointed to a private server. I do want to iterate in the workflow pipelines in the near future. We can look into adding it back, but I don't think it's a blocker right now.

File: src/main/java/fr/formiko/mc/underilla/paper/cleaning/CleanBlocks.java
Comment (kntrel) [reply src/main/java/fr/formiko/mc/underilla/paper/cleaning/CleanBlocks.java:1]: I do agreee. But, if I read the code correctly, it was part of the already depricated worl-wile cliean up task. So, if I'm not wrong, it wasn't firing anyway.

If my assertion is true. I think we can leave it removed in this version, because it woudln't break previous behavior. If I'm wrong, this is a merge blocker indeed.

File: README.md (hunk 1 above)
Comment (kntrel) [reply README.md:10]: Let's take this offline :)

File: src/main/java/fr/formiko/mc/underilla/paper/cleaning/CleanBlocks.java
Comment (kntrel) [reply src/main/java/fr/formiko/mc/underilla/paper/cleaning/CleanBlocks.java:1]: Nervermind. Added the functionality back in [5a6fea0](https://github.com/kntrel/Underilla/pull/18/commits/5a6fea063ef9c193236b56834ed3b6d18c212ca8).

File: .github/scripts/get-md-section.py
Comment (HydrolienF) [inline .github/scripts/get-md-section.py:1]: I think it would be better to have it handle by gradle plugin.

**Review by HydrolienF — APPROVED** [review]
Overall the update looks good to me. I've send comment on all the small changes I think deserved some change but most of it was improvement.

File: .github/scripts/get-md-section.py
Comment (kntrel) [reply .github/scripts/get-md-section.py:1]: Fair enough. We can look into this offline and make a desicions for further refining
